mesonbuild/modules/pkgconfig.py

```diff
@@ -374,6 +375,16 @@ def _generate_pkgconfig_file(self, state, deps, subdirs, name, description,
         else:
             outdir = state.environment.scratch_dir
             prefix = PurePath(coredata.get_option(mesonlib.OptionKey('prefix')))
+            if pkgroot:
+                pkgroot = PurePath(pkgroot)
+                if pkgroot.is_absolute():
+                    try:
+                        pkgroot = pkgroot.relative_to(prefix)
+                    except ValueError:
+                        raise mesonlib.MesonException('Pkgconfig prefix cannot be outside of the prefix '
+                                                      'when pkgconfig.relocatable=true. '
+                                                      f'Pkgconfig prefix is {pkgroot}.')
+                prefix = PurePath('${pcfiledir}', *('..' for _ in pkgroot.parts))
```

Comment:
```python
if not pkgroot.is_absolute():
    pkgroot = prefix / pkgroot
elif prefix not in pkgroot.parents:
    raise mesonlib.MesonException('...')
prefix = PurePath('${pcfiledir}', os.path.relpath(prefix, pkgroot))
```

The use of one `'..'` per part in pkgroot feels wrong to me, but PurePath().relative_to() is broken by design as it doesn't allow you to compute the relative position of two absolute paths. This is one more example of how pathlib's best feature is introducing an os.PathLike interface, but their actual implementation decisions are... often bad.

Fortunately, the os.path.* family of methods still work fine, and they accept PathLike objects. In this case, `os.path.relpath` > `PurePath.relative_to`.

(Don't let me get started on how utterly catastrophic the entire set of concrete Path() methods are, with their insistence that .exists() should raise a traceback if a path cannot exist, or doesn't seem to exist, due to stat errors... but it's totally fine to check if an invalid pathname exists, that just immediately returns `False` lolwut.)

Reply:
Yeah, it could be fragile if `pkgroot` for some reason contains `'..'`. The `os.path.relpath` implementation looks a lot cleaner. I'll be switching to that. And I should have listen to @xclaesse.

Thanks for the input.

Reply:
Should be all good now.